### PR TITLE
2204 curves

### PR DESCRIPTION
**Pull request by LiamGallagherBC**
Please merge before: https://github.com/BeatConnect/bc_unity_daw/pull/2233

# Overview
The goal of this PR is to support additional curve types to be used with macro controls. For more details and demos see the bc_unity_daw PR linked above.

### CODE REVIEW

File: modules/tracktion_engine/model/automation/tracktion_AutomatableParameter.cpp [hunk 1]
```diff
@@ -317,11 +364,19 @@ struct MacroSource : public AutomationModifierSource
...
         auto macroValue = macro->getCurrentValue();
+        
+        // BEATCONNECT MODIFICATION START
+        AutomationScaleHelpers::CurveType curveType;
 
-        const auto range = juce::Range<float>::between (assignment->inputLimitStart.get(), assignment->inputLimitEnd.get());
-        currentValue.store (AutomationScaleHelpers::mapValue (AutomationScaleHelpers::limitInputValue (macroValue, range),
-                                                              assignment->offset, assignment->value, assignment->curve),
-                            std::memory_order_release);
+        if (!assignment->state.hasProperty("curveType"))
```
Comment (BenoitCharraBC) [inline modules/tracktion_engine/model/automation/tracktion_AutomatableParameter.cpp]: Where is "curveType" set? Somewhere externally?

Comment (LiamGallagherBC) [reply modules/tracktion_engine/model/automation/tracktion_AutomatableParameter.cpp]: That's correct, it's set by right clicking on the curve control on the macro building and then selecting one of the four options. 
image

Comment (BenoitCharraBC) [reply modules/tracktion_engine/model/automation/tracktion_AutomatableParameter.cpp]: OK. Just put "curveType" into IDs in tracktion_Identifiers.h with a BEATCONNECT MODIFICATION.
Maybe here:

image

I don't see how a tracktion modification is avoidable in this case. That said, have you fully investigated this?

Comment (LiamGallagherBC) [reply modules/tracktion_engine/model/automation/tracktion_AutomatableParameter.cpp]: The two scenarios I considered for applying the macro curve were:
1) Do the math within a property change ahead of tracktion engine being aware of the curve, and then setting the curve value to 0.
2) Have plugins check their automatable parameters and redo the curve math there.

These approaches would leave tracktion engine unchanged, but also they would mean that the curves wouldn't be accounted for in the automation system, and that we would end up with two parallel systems for calculating curves.

I think these changes are made in the right place, where tracktion does all of its curved macro parameter value math; and minimal, where for 3 of the cases I was able to largely reuse existing code. It's not no changes to the engine but I feel they're reasonable.

**Review by BenoitCharraBC — APPROVED** [review]

**Pull request merged**